### PR TITLE
gh-1820 revise create_index() page

### PR DESCRIPTION
**Pull request by Onvember**

### CODE REVIEW

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 1]
```diff
@@ -18,13 +17,13 @@ On this page:
...
 
         Create an :ref:`index <index-box_index>`.
-        It is mandatory to create an index for a space
-        before trying to insert tuples into it, or select tuples from it. The
-        first created index, which will be used as the primary-key index, must be
-        unique.
+
+        It is mandatory to create an index for a space before trying to insert
+        tuples into it, or select tuples from it. The first created index, which
+        will be used as the primary-key index, must be unique.
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:26]: Can we straighten the logic a bit?

```suggestion
        tuples into it, or select tuples from it. The first created index
        will be used as the primary-key index, so it must be unique.
```

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 2]
```diff
@@ -36,6 +35,18 @@ On this page:
...
+        **Possible errors:**
+
+        * too many parts;
+        * index '...' already exists;
+        * primary key must be unique.
+
+        Building or rebuilding a large index will cause occasional
+        :ref:`yields <atomic-cooperative_multitasking>`
+        so that other requests will not be blocked.
+        If the other requests cause an illegal situation such as a duplicate key
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:47]: Is "illegal situation" a term? Do we define it anywhere?

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 3]
```diff
@@ -36,6 +35,18 @@ On this page:
...
+        in a unique index, the index building or rebuilding will fail.
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:48]:
```suggestion
        in a unique index, building or rebuilding such index will fail.
```

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 4]
```diff
@@ -98,8 +109,8 @@ On this page:
             |                     | faster, ``false`` -- an index size is reduced by half |                                  |                               |
             +---------------------+-------------------------------------------------------+----------------------------------+-------------------------------+
 
-            The options in the above chart are also applicable for
-            :doc:`/reference/reference_lua/box_index/alter`.
+        The options in the above chart are also applicable for
+        :doc:`/reference/reference_lua/box_index/alter`.
 
 
         **Note re storage engine:** vinyl has extra options which by default are
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:116]: "re" is not so popular :)

```suggestion
        **Note about storage engine:** vinyl has extra options which by default are
```

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 5]
```diff
@@ -113,321 +124,404 @@ On this page:
...
-
-        **Possible errors:**
-
-        * too many parts;
-        * index '...' already exists;
-        * primary key must be unique.
+        **Example:**
 
-        .. code-block:: tarantoolsession
+        ..  code-block:: tarantoolsession
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:129]: nice )

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 6]
```diff
@@ -113,321 +124,404 @@ On this page:
...
+
+Index field types differ depending on what values are allowed,
+and what index types are allowed.
+
+..  container:: table stackcolumn
+
+    ..  rst-class:: left-align-column-1
+    ..  rst-class:: left-align-column-2
+
+    ..  list-table::
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:169]: A table is cool )

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 7]
```diff
@@ -113,321 +124,404 @@ On this page:
...
+        :widths: 10 45 20 15
+        :header-rows: 1
+
+        *   - Index field type
+            - What can be it it
+            - Where it is legal
+            - Examples
+
+        *   - **unsigned**
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:178]: Looks like code to me, should be formatted as code.

```suggestion
        *   - ``unsigned``
```

File: doc/reference/reference_lua/box_space/create_index.rst (hunk 4 above)
Comment (Onvember) [reply doc/reference/reference_lua/box_space/create_index.rst:116]: We have this phrase declared here https://www.tarantool.io/en/doc/latest/book/box/engines/
So if you want to change the wording, let's do it in another issue.

Comment (NickVolynkin) [reply doc/reference/reference_lua/box_space/create_index.rst:116]: Hmm, okay.

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 8]
```diff
@@ -113,321 +124,404 @@ On this page:
...
+            -
+            -
+
+.. _box_space-is_nullable:
+
+--------------------------------------------------------------------------------
+Allowing null for an indexed key
+--------------------------------------------------------------------------------
+
+If the index type is TREE, and the index is not the primary index,
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:305]: TREE should be formatted as inline code.

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 9]
```diff
@@ -113,321 +124,404 @@ On this page:
...
+then the ``parts={...}`` clause may include ``is_nullable=true`` or
+``is_nullable=false`` (the default). If ``is_nullable`` is true,
+then it is legal to insert ``nil`` or an equivalent such as ``msgpack.NULL``
+(or it is legal to insert nothing at all for trailing nullable fields).
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:309]: parentheses → new sentence

```suggestion
then it is legal to insert ``nil`` or an equivalent such as ``msgpack.NULL``.
It is also legal to insert nothing at all when using trailing nullable fields.
```

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 10]
```diff
@@ -113,321 +124,404 @@ On this page:
...
+for JSON documents that all have the same structure.
+
+For example, when creating an index on field#2 for a string document
+that will start with ``{'data': [{'name': '...'}, {'name': '...'}]``,
+the parts section in the create_index request could look like:
+``parts = {{field = 2, type = 'str', path = 'data[*].name'}}``.
+Then tuples containing names can be retrieved quickly with
+``index_object:select({key-value})``.
+
+In fact a single field can have multiple keys, as in this example
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:410]: 1. In fact, a single field (comma)
2. The sentence is too complex. Try breaking it into two or three separate sentences.

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 11]
```diff
@@ -113,321 +124,404 @@ On this page:
...
+Functional indexes are indexes that call a user-defined function for forming
+the index key, rather than depending entirely on the Tarantool default formation.
+Functional indexes are useful for condensing or truncating or reversing or
+any other way that users want to customize the index.
+
+The function definition must expect a tuple (which has the contents of
+fields at the time a data-change request happens) and must return a tuple
+(which has the contents that will actually be put in the index).
+
+The space must have a memtx engine.
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:464]: This section has a lot of one-sentence paragraphs. Maybe they can be joined into a few longer paragraphs or grouped in a list.

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 12]
```diff
@@ -113,321 +124,404 @@ On this page:
...
         box.schema.space.create('x', {engine = 'memtx'})
         box.space.x:create_index('i',{parts={{field = 1, type = 'string'}}})
-        -- Step 2: Make the function.
-        -- The function expects a tuple. In this example it will work on tuple[2]
-        -- because the key source is field number 2 in what we will insert.
-        -- Use string.sub() from the string module to get the first character.
+
+#.  Make the function: the function expects a tuple. In this example it will
+    work on tuple[2] because the key source is field number 2 in what we will
+    insert. Use string.sub() from the string module to get the first character.
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:499]:
```suggestion
    insert. Use ``string.sub()`` from the ``string`` module to get the first character:
```

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 13]
```diff
@@ -113,321 +124,404 @@ On this page:
...
+
+    ..  code-block:: lua
+
         lua_code = [[function(tuple) return {string.sub(tuple[2],1,1)} end]]
-        -- Step 3: Make the function persistent.
-        -- Use the box.schema.func.create function for this.
+
+#.  Make the function persistent: use the box.schema.func.create function for this.
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:505]: Either formatting or a link to the document.
```suggestion
#.  Make the function persistent using the ``box.schema.func.create`` function:
```

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 14]
```diff
@@ -113,321 +124,404 @@ On this page:
...
+
+    ..  code-block:: lua
+
         box.schema.func.create('F',
             {body = lua_code, is_deterministic = true, is_sandboxed = true})
-        -- Step 4: Make the functional index.
-        -- Specify the fields whose values will be passed to the function.
-        -- Specify the function.
+
+#.  Make the functional index: specify the fields whose values will be passed
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:512]: This is the first mention of this index, so we should use "a": "Make a functional index".

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 15]
```diff
@@ -113,321 +124,404 @@ On this page:
...
+
         box.space.x:create_index('j',{parts={{field = 1, type = 'string'}},func = 'F'})
-        -- Step 5: Test.
-        -- Insert a few tuples.
-        -- Select using only the first letter, it will work because that is the key
-        -- Or, select using the same function as was used for insertion
+
+#.  Test. Insert a few tuples. Select using only the first letter, it will work
+    because that is the key. Or, select using the same function as was used for
+    insertion
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:521]:
```suggestion
    insertion:
```

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 16]
```diff
@@ -113,321 +124,404 @@ On this page:
...
+    ..  code-block:: lua
+
         box.schema.func.create('F',
             {body = lua_code, is_deterministic = true, is_sandboxed = true})
-        -- Step 4: Make the functional index.
-        -- Specify the fields whose values will be passed to the function.
-        -- Specify the function.
+
+#.  Make the functional index: specify the fields whose values will be passed
+    to the function. Specify the function.
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:513]: We're giving an example after this sentence.

```suggestion
    to the function. Specify the function:
```

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 17]
```diff
@@ -446,38 +540,54 @@ On this page:
...
+    lua_code = [[function(tuple) return {string.sub(tuple[2],1,1)} end]]
+    box.schema.func.create('F',
+        {body = lua_code, is_deterministic = true, is_sandboxed = true})
+    box.space.x:create_index('j',{parts={{field = 1, type = 'string'}},func = 'F'})
+    box.space.x:insert{'a', 'wombat'}
+    box.space.x:insert{'b', 'rabbit'}
+    box.space.x.index.j:select('w')
+    box.space.x.index.j:select(box.func.F:call({{'x', 'wombat'}}));
+
+Functions for functional indexes can return **multiple keys**. Such functions are
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:558]: Why bold?

File: doc/reference/reference_lua/box_space/create_index.rst [hunk 18]
```diff
@@ -113,321 +124,404 @@ On this page:
...
+
+Functional indexes must not be primary-key indexes.
+
+Functional indexes cannot be altered and the function cannot be changed if
+it is used for an index, so the only way to change them is to drop the index
+and create it again.
+
+Only sandboxed functions are suitable for functional indexes.
+
+**Example:**
```
Comment (NickVolynkin) [inline doc/reference/reference_lua/box_space/create_index.rst:485]: Why bold and not a heading?

**Review by NickVolynkin — CHANGES_REQUESTED** [review]

**Review by NickVolynkin — APPROVED** [review]

File: doc/reference/reference_lua/box_space/create_index.rst (hunk 11 above)
Comment (NickVolynkin) [reply doc/reference/reference_lua/box_space/create_index.rst:464]: > в [*]-секции хотели сделать маркированный список, но не получилось, надо исправить

**Pull request merged**